Added AWS File system support

TimurKustov opened the pull request.

AlekseiLitkovetc committed:
S3 permissions errors using Cromwell 37+ BA-4740 (#4740)

Lumen06 committed:
Add AWS File system support;
Add a unit test for AwsFileChecker

Lumen06 committed:
Add AWS File system support;
Add a unit test for AwsFileChecker

SergeySdv committed:
Add Aws credentials section in centaur's config
Add unit tests

SergeySdv committed:
Add Aws credentials section in centaur's config
Add unit tests

DenisVerkhoturov commented on centaur/src/main/resources/reference.conf:
```diff
@@ -75,5 +75,12 @@ centaur {
+  access-key = ""
+  secret-key = ""
+  region = ""
+  }
```
Format it, please, there is supposed to be an indentation in the `aws` block.
```diff
   aws {
-  application-name = "centaur"
-  access-key = ""
-  secret-key = ""
-  region = ""
+    application-name = "centaur"
+    access-key = ""
+    secret-key = ""
+    region = ""
   }
```

iliaonishchenko commented on centaur/src/main/scala/centaur/test/FilesChecker.scala:
```diff
@@ -0,0 +1,95 @@
+case class AwsS3Ops(amazonS3: AmazonS3) {
+
+  def parsePath: String => AwsS3Path = { fullPath =>
+    val bucketAndDashes = fullPath.drop(5).split('/')
```
You can use another `split` method, which is `split(regex, limit)`
In other words, `split("/", 2)` will do the job.

DenisVerkhoturov replied:
What does this magic `5` mean?

The same in the `GCSOps` class.

iliaonishchenko commented on centaur/src/main/scala/centaur/test/FilesChecker.scala:
```diff
@@ -0,0 +1,95 @@
+
+case object LocalFilesChecker extends FilesChecker {
+  def countObjectsAtPath: String => Int = { s =>
+    val d = new java.io.File(s)
```
Please import `File` and not declare path to class in the code.

iliaonishchenko commented on centaur/src/main/scala/centaur/test/FilesChecker.scala:
```diff
@@ -0,0 +1,95 @@
+}
+
+object AwsS3 {
+  implicit def awsOps(amazonS3: AmazonS3): AwsS3Ops = AwsS3Ops(amazonS3)
```
Implicit conversions is evil.

iliaonishchenko replied:
Consider type classes instead of implicit conversions.

iliaonishchenko commented on centaur/src/main/scala/centaur/test/FilesChecker.scala:
```diff
@@ -0,0 +1,95 @@
+    storage.parsePath andThen storage.countObjectsAtPath
+}
+
+case object LocalFilesChecker extends FilesChecker {
```
Why it is `case object` instead of object?

iliaonishchenko commented on centaur/src/main/scala/centaur/test/Test.scala:
```diff
@@ -119,6 +122,18 @@ object Operations {
     storageOptions.getService
   }
 
+  lazy val awsConfiguration: AwsConfiguration = AwsConfiguration(CentaurConfig.conf)
```
Why lazy?
Is there a chance it won't be used?

SergeySdv replied:
I suppose if we will not use aws s3 options for checking. For example we coud use only local file checking or gs.

DenisVerkhoturov commented on project/Dependencies.scala:
```diff
@@ -262,6 +263,7 @@ object Dependencies {
 
   private val awsCloudDependencies = List(
     "com.fasterxml.jackson.core" % "jackson-annotations" % jacksonV,
+    "com.amazonaws" % "aws-java-sdk-s3" % amazonAwsV
```
This dependency looks suspiciously alike the existing one [261: `"software.amazon.awssdk" % "s3" % awsSdkV`](https://github.com/broadinstitute/cromwell/blob/develop/project/Dependencies.scala#L261). Are you sure we need another one?

iliaonishchenko commented on centaur/src/main/scala/centaur/test/Test.scala:
```diff
@@ -119,6 +122,18 @@ object Operations {
+
+  def buildAmazonS3Client: AmazonS3 = {
+    val basicAWSCredentials = new BasicAWSCredentials(accessKeyId, secretAccessKey)
+    AmazonS3ClientBuilder.standard.withCredentials(new AWSStaticCredentialsProvider(basicAWSCredentials)).withRegion(region).build
```
Please consider doing this not in one line.

iliaonishchenko commented on centaur/src/test/scala/centaur/s3filechecking/NegTestAwsFileChecker.scala:
```diff
@@ -0,0 +1,46 @@
+import org.scalatest.FlatSpec
+import org.specs2.matcher.ShouldMatchers
+
+class NegTestAwsFileChecker extends FlatSpec with ShouldMatchers with MockFactory {
```
Class name does not look like it is a test.

DenisVerkhoturov replied:
Please use `spec2.Mokito` since it is already defined as a dependency.
```scala
import org.specs2.mock.Mockito
import org.mockito.Mockito._

class NegTestAwsFileChecker extends FlatSpec with ShouldMatchers with Mockito { /**/ }
```

iliaonishchenko commented on centaur/src/test/scala/centaur/s3filechecking/TestAwsFileChecker.scala:
```diff
@@ -0,0 +1,50 @@
+import org.scalatest.FlatSpec
+import org.specs2.matcher.ShouldMatchers
+
+class TestAwsFileChecker extends FlatSpec with ShouldMatchers with MockFactory {
```
Why we have two test classes testing the same functionality?

DenisVerkhoturov replied:
Please, merge these two specifications (`NegTestAwsFileChecker` and `TestAwsFileChecker`) into a single one with all the negative and positive cases.

In addition, all the specifications usually are named by the following: the name of the entity that is tested plus **Spec** suffix (for instance: `ListSpec`, `UserControllerSpec`, etc).

iliaonishchenko reviewed (changes requested):
Nice PR.
Changes is required.

iliaonishchenko commented:
By the way, @Lumen06 @TimurKustov are you working as two developers? :)

DenisVerkhoturov commented on centaur/src/test/scala/centaur/s3filechecking/NegTestAwsFileChecker.scala:
```diff
@@ -0,0 +1,46 @@
+}
+
+
+
```
Please, clean up these empty lines.

Lumen06 committed:
Refactor: ObjectCounter, FileChecker and unit test FileCheckerSpec

Common logic is defined at parsePath at ObjectCounter now. Also parsePath takes as argument a regex for specify filechecker, which are at FileCheckers as private fields. FileCheckerSpec moved to centaur/src/test/scala/centaur/testfilecheck

Lumen06 committed:
Refactor: ObjectCounter, FileChecker and unit test FileCheckerSpec

Common logic is defined at parsePath at ObjectCounter now. Also parsePath takes as argument a regex for specify filechecker, which are at FileCheckers as private fields. FileCheckerSpec moved to centaur/src/test/scala/centaur/testfilecheck

DenisVerkhoturov commented on centaur/src/test/scala/centaur/s3filechecking/NegTestAwsFileChecker.scala:
```diff
@@ -0,0 +1,46 @@
+    (amazonS3Stub.listObjects(_: String, _: String)).when(bucketName, dirName).returns(objectListing)
+    val awsS3Path = awsS3Ops.parsePath(wrongTestPath)
+    assert(awsS3Path.bucket != bucketName)
+    assert(awsS3Path.directory != dirName)
```
`assert`-s are not informative at all when things come to investigating the reasons for failing tests, because they are only able to say that some boolean is not true. It is always great to the exact values. So, please, update your tests using matchers. Example:
```scala
awsS3Path.bucket shouldNotEqual bucketName
```

DenisVerkhoturov commented on centaur/src/test/scala/centaur/s3filechecking/NegTestAwsFileChecker.scala:
```diff
@@ -0,0 +1,46 @@
+
+    objectListing.setBucketName(bucketName)
+    objectListing.setPrefix(dirName)
+    (amazonS3Stub.listObjects(_: String, _: String)).when(bucketName, dirName).returns(objectListing)
```
In the scope of the method that is tested here (`parsePath`) `listObjects` is never called. So you don't need to define this behaviour.

DenisVerkhoturov reviewed (changes requested):
Great work here! 👍 
Thank you, here you can find some thoughts.

General recommendations:
Please, consider rewriting this logic using the `awssdk` that is already defined in the project, no need to define another one (`aws-java-sdk-s3`).

SergeySdv committed:
Add Negative scenario unit tests for FilesChecker.scala

svitkovsergey commented on centaur/src/main/scala/centaur/test/ObjectCounter.scala:
```diff
@@ -0,0 +1,51 @@
+      val prefixLength = 5
+      val bucketAndDashes = fullPath.drop(prefixLength).split("/", 2)
+      val bucket = bucketAndDashes.head
+      val directory = bucketAndDashes.tail.mkString("/")
```
Since we are now using `split(..., 2)`, there is no need to use `tail.mkString("/")`

Lumen06 committed:
refactor: ObjectCounter, FileCheckSpec, Test.scala, Dependencies

Added new client at Test.scala, refactored parsePath at ObjectCounter, FileCheckSpec rewrited with Mockito, now used "software.amazon.awssdk" % "s3" % awsSdkV instead of "com.amazonaws" % "aws-java-sdk-s3" % amazonAwsV"

svitkovsergey reviewed (changes requested):
Good request!

SergeySdv committed:
Add Negative scenario unit tests for FilesChecker.scala

Commit:
Change config structure

svitkovsergey commented:
Not actual anymore.
